lib/charms/prometheus_pushgateway_k8s/v0/pushgateway.py

```diff
@@ -0,0 +1,172 @@
+
+```shell
+cd some-charm
+charmcraft fetch-lib charms.prometheus_pushgateway_k8s.v0.pushgateway
+```
+
+In the `metadata.yaml` of the charm, add the following:
+
+```yaml
+requires:
+  your-relation-name:
+    interface: pushgateway
+```
+
+In the source of your charm, first import the interface:
+
+```
+from charms.prometheus_pushgateway_k8s.v0.pushgateway import PrometheusPushgatewayRequirer
+````
+
+Instantiate the object in your charm's `__init__`, like so:
+
+```
+from charms.prometheus_pushgateway_k8s.v0.pushgateway import PrometheusPushgatewayRequirer
+from ops.charm import CharmBase
+
+class MyCharm(CharmBase):
+    def __init__(...):
+         ...
+        self.pushgateway_requirer = PrometheusPushgatewayRequirer(self, "your-relation-name")
+```
+
+Then use it at any moment to send a metric (validating that the requirer is ready), passing
+its name and value:
+
+```
+    if self.pushgateway_requirer.is_ready():
+        self.pushgateway_requirer.send_metric("test_metric", 3.141592)
+```
+
+The requirer is ready when the relation to the Prometheus Pushgateway is properly established.
+
+The `send_metric` call will just end quietly if the metric was sent succesfully, or will raise
+an exception if something is wrong (that error should be logged or informed to the operator).
+"""
+
+import json
+import logging
+import socket
+from typing import Optional, Union
+from urllib import request
+from urllib.error import HTTPError
+
+from ops.charm import CharmBase, RelationEvent
+from ops.framework import Object
+
+logger = logging.getLogger(__name__)
+
+
+# The unique Charmhub library identifier, never change it
+LIBID = "a0065690fe484ef296d5847fcbf1d728"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+# the key in the relation data
+RELATION_KEY = "push-endpoint"
+
+
+class PrometheusPushgatewayProvider(Object):
+    """Provider side for the Prometheus Pushgateway.
+
+    This class is to be used by the Prometheus Pushgateway charm, please use
+    the PrometheusPushgatewayRequirer class if you're bulding a charm and want to
+    use this library to integrate with the Prometheus Pushgateway.
+    """
+
+    def __init__(self, charm: CharmBase, relation_name: str, port: int):
+        super().__init__(charm, relation_name)
+        self.port = port
+        self.app = charm.app
+        events = charm.on[relation_name]
+        self.framework.observe(events.relation_created, self._on_relation_changed)
+        self.framework.observe(events.relation_changed, self._on_relation_changed)
+
+    def _on_relation_changed(self, event: RelationEvent):
+        """Send the push endpoint info."""
+        relation_data = event.relation.data[self.app]
+        relation_data[RELATION_KEY] = json.dumps(
+            {
+                "hostname": socket.getfqdn(),
```

Comment:
For typical deployment scenarios of push gateway, would this ever need to happen through an ingress?

Reply:
I followed what I saw in other libraries here. We can revisit in a later PR, if we consistently move to other solution.